velox/exec/GroupingSet.cpp

```diff
@@ -412,4 +444,241 @@ const HashLookup& GroupingSet::hashLookup() const {
   return *lookup_;
 }
 
+namespace {
+// Checks if it is likely that the reservation on 'tracker' can be
+// incremented by 'increment'. Returns false if this seems
+// unlikely. Otherwise attempts the reservation increment and
+// returns true if succeeded. This will be moved to Task when adding
+// memory contention arbitration.
+bool maybeReserve(int64_t increment, memory::MemoryUsageTracker& tracker) {
```

Comment:
I feel that maybeReserve logic deserves a test. Like you mentioned it doesn't belong to GroupingSet to begin this. Consider extracting it into a separate PR.